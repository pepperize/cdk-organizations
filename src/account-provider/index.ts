export * from "./account-provider";
export * from "./is-complete-handler-function";
export * from "./on-event-handler-function";
