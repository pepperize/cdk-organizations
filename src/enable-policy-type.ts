import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId } from "aws-cdk-lib/custom-resources";
import { Construct } from "constructs";
import { Root } from "./organization";
import { PolicyType } from "./policy";

export interface EnablePolicyTypeProps {
  readonly root: Root;
  readonly policyType: PolicyType;
}

/**
 * Enables and disables Enables a policy type in a root. After you enable a policy type in a root, you can attach policies of that type to the root, any organizational unit (OU), or account in that root.
 *
 * @see https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies_enable-disable.html
 */
export class EnablePolicyType extends Construct {
  public constructor(scope: Construct, id: string, props: EnablePolicyTypeProps) {
    super(scope, id);

    const { root, policyType } = props;

    new AwsCustomResource(this, "EnablePolicyTypeCustomResource", {
      resourceType: "Custom::Organizations_EnablePolicyType",
      onCreate: {
        service: "Organizations",
        action: "enablePolicyType", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#enablePolicyType-property
        region: "us-east-1",
        physicalResourceId: PhysicalResourceId.of(`${root.rootId}:${policyType}`),
        parameters: {
          RootId: root.rootId,
          PolicyType: policyType,
        },
      },
      onDelete: {
        service: "Organizations",
        action: "disablePolicyType", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#disablePolicyType-property
        region: "us-east-1",
        parameters: {
          RootId: root.rootId,
          PolicyType: policyType,
        },
      },
      installLatestAwsSdk: false,
      policy: AwsCustomResourcePolicy.fromSdkCalls({
        resources: AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });
  }
}
