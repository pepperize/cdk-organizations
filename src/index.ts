export * from "./account";
export * from "./organization";
export * from "./organizational-unit";
export * from "./parent";
export * from "./policy";
export * from "./policy-attachment";
