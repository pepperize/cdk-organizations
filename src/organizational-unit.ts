import { Annotations, CustomResource, RemovalPolicy, TagManager, TagType } from "aws-cdk-lib";
import { Construct, IConstruct } from "constructs";
import { OrganizationalUnitProvider } from "./organizational-unit-provider/organizational-unit-provider";
import { IChild, IParent } from "./parent";
import { IPolicyAttachmentTarget } from "./policy-attachment";
import { ITaggableResource, TagResource } from "./tag-resource";
import { Validators } from "./validators";

export interface OrganizationalUnitProps {
  /**
   * The friendly name to assign to the new OU.
   */
  readonly organizationalUnitName: string;
  /**
   * The parent root or OU that you want to create the new OrganizationalUnit in.
   */
  readonly parent: IParent;
  /**
   * Whether to import, if a duplicate organizational unit with same name exists in the parent exists.
   *
   * @default true
   */
  readonly importOnDuplicate?: boolean;
  /**
   * If set to RemovalPolicy.DESTROY, the organizational unit will be deleted
   *
   * @default RemovalPolicy.Retain
   */
  readonly removalPolicy?: RemovalPolicy;
}

/**
 * A container for accounts within a root. An OU also can contain other OUs, enabling you to create a hierarchy that resembles an upside-down tree, with a root at the top and branches of OUs that reach down, ending in accounts that are the leaves of the tree. When you attach a policy to one of the nodes in the hierarchy, it flows down and affects all the branches (OUs) and leaves (accounts) beneath it. An OU can have exactly one parent, and currently each account can be a member of exactly one OU.
 *
 * <strong>You must first move all accounts out of the OU and any child OUs, and then you can delete the child OUs.</strong>
 */
export interface IOrganizationalUnit extends IPolicyAttachmentTarget, IParent, IChild, IConstruct {
  /**
   * The unique identifier (ID) associated with this OU. The regex pattern for an organizational unit ID string requires "ou-" followed by from 4 to 32 lowercase letters or digits (the ID of the root that contains the OU). This string is followed by a second "-" dash and from 8 to 32 additional lowercase letters or digits.
   */
  readonly organizationalUnitId: string;
  /**
   * The Amazon Resource Name (ARN) of this OU. For more information about ARNs in Organizations, see [ARN Formats Supported by Organizations](https://docs.aws.amazon.com/service-authorization/latest/reference/list_awsorganizations.html#awsorganizations-resources-for-iam-policies) in the AWS Service Authorization Reference.
   */
  readonly organizationalUnitArn: string;
  /**
   * The friendly name of this OU.
   */
  readonly organizationalUnitName: string;
}

export class OrganizationalUnit extends Construct implements IOrganizationalUnit, ITaggableResource {
  readonly organizationalUnitId: string;
  readonly organizationalUnitArn: string;
  readonly organizationalUnitName: string;
  readonly tags = new TagManager(TagType.KEY_VALUE, "Custom::Organizations_OrganizationalUnitProvider");

  public constructor(scope: Construct, id: string, props: OrganizationalUnitProps) {
    super(scope, id);

    const { organizationalUnitName, parent, importOnDuplicate, removalPolicy } = props;

    if (!Validators.of().organizationalUnitName(organizationalUnitName)) {
      Annotations.of(this).addError(
        "The organizational unit's name must be of type string and between 1 and 128 characters long."
      );
    }

    this.node.addDependency(parent);

    const organizationalUnitProvider = OrganizationalUnitProvider.getOrCreate(this);
    const organizationalUnit = new CustomResource(this, "OrganizationProvider", {
      serviceToken: organizationalUnitProvider.provider.serviceToken,
      resourceType: "Custom::Organizations_OrganizationalUnitProvider",
      properties: {
        Name: organizationalUnitName,
        ParentId: parent.identifier(),
        ImportOnDuplicate: String(importOnDuplicate ?? true),
        RemovalPolicy: removalPolicy ?? RemovalPolicy.RETAIN,
      },
    });

    this.organizationalUnitId = organizationalUnit.getAtt("Id").toString();
    this.organizationalUnitArn = organizationalUnit.getAtt("Arn").toString();
    this.organizationalUnitName = organizationalUnit.getAtt("Name").toString();

    const tagResource = new TagResource(this, "Tags", { resource: this });
    tagResource.node.addDependency(organizationalUnit);
  }

  identifier(): string {
    return this.organizationalUnitId;
  }
}
