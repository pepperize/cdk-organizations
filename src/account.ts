import { Annotations, CustomResource, RemovalPolicy, TagManager, TagType } from "aws-cdk-lib";
import { Construct, IConstruct } from "constructs";
import { pascalCase } from "pascal-case";
import { AccountProvider } from "./account-provider";
import { DelegatedAdministrator } from "./delegated-administrator";
import { IChild, IParent } from "./parent";
import { IPolicyAttachmentTarget } from "./policy-attachment";
import { IResource } from "./resource";
import { ITaggableResource, TagResource } from "./tag-resource";
import { Validators } from "./validators";

/**
 * @see https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/control-access-billing.html#ControllingAccessWebsite-Activate
 */
export enum IamUserAccessToBilling {
  /**
   * If set to ALLOW, the new account enables IAM users to access account billing information if they have the required permissions.
   */
  ALLOW = "ALLOW",
  /**
   * If set to DENY, only the root user of the new account can access account billing information.
   */
  DENY = "DENY",
}

export interface AccountProps {
  /**
   * The email address of the owner to assign to the new member account. This email address must not already be associated with another AWS account. You must use a valid email address to complete account creation. You can't access the root user of the account or remove an account that was created with an invalid email address.
   */
  readonly email: string;
  /**
   * The friendly name of the member account.
   */
  readonly accountName: string;
  /**
   * The name of an IAM role that AWS Organizations automatically preconfigures in the new member account. This role trusts the management account, allowing users in the management account to assume the role, as permitted by the management account administrator. The role has administrator permissions in the new member account.
   *
   * If you don't specify this parameter, the role name defaults to OrganizationAccountAccessRole.
   */
  readonly roleName?: string;
  /**
   * If set to ALLOW , the new account enables IAM users to access account billing information if they have the required permissions. If set to DENY , only the root user of the new account can access account billing information.
   *
   * @default ALLOW
   */
  readonly iamUserAccessToBilling?: IamUserAccessToBilling;

  /**
   * The parent root or OU that you want to create the new Account in.
   */
  readonly parent?: IParent;
  /**
   * Whether to import, if a duplicate account with same name and email already exists.
   *
   * @default true
   */
  readonly importOnDuplicate?: boolean;
  /**
   * If set to RemovalPolicy.DESTROY, the account will be moved to the root.
   *
   * @default RemovalPolicy.Retain
   */
  readonly removalPolicy?: RemovalPolicy;
}

export interface IAccount extends IPolicyAttachmentTarget, IChild, IConstruct, IResource {
  /**
   * If the account was created successfully, the unique identifier (ID) of the new account. Exactly 12 digits.
   */
  readonly accountId: string;
  /**
   * The Amazon Resource Name (ARN) of the account.
   */
  readonly accountArn: string;
  /**
   * The friendly name of the account.
   */
  readonly accountName: string;
  /**
   * The email address of the owner to assign to the new member account. This email address must not already be associated with another AWS account. You must use a valid email address to complete account creation. You can't access the root user of the account or remove an account that was created with an invalid email address.
   */
  readonly email: string;
}

/**
 * Creates or imports an AWS account that is automatically a member of the organization whose credentials made the request. AWS Organizations automatically copies the information from the management account to the new member account
 */
export class Account extends Construct implements IAccount, ITaggableResource {
  public readonly accountId: string;
  public readonly accountArn: string;
  public readonly accountName: string;
  public readonly email: string;

  protected readonly resource: CustomResource;

  readonly tags = new TagManager(TagType.KEY_VALUE, "Custom::Organizations_Account");

  public constructor(scope: Construct, id: string, props: AccountProps) {
    super(scope, id);

    const { email, accountName, roleName, iamUserAccessToBilling, parent, importOnDuplicate, removalPolicy } = props;

    if (!Validators.of().email(email)) {
      Annotations.of(this).addError("The account's email must be of type string and between 6 and 64 characters long.");
    }

    if (!Validators.of().accountName(accountName)) {
      Annotations.of(this).addError("The account's name must be of type string and between 1 and 50 characters long.");
    }

    const createAccountProvider = AccountProvider.getOrCreate(this);
    const account = new CustomResource(this, "CreateAccount", {
      serviceToken: createAccountProvider.provider.serviceToken,
      resourceType: "Custom::Organizations_Account",
      properties: {
        Email: email,
        AccountName: accountName,
        RoleName: roleName ?? "OrganizationAccountAccessRole",
        IamUserAccessToBilling: iamUserAccessToBilling ?? IamUserAccessToBilling.ALLOW,
        ParentId: parent?.identifier(),
        ImportOnDuplicate: String(importOnDuplicate ?? true),
        RemovalPolicy: removalPolicy ?? RemovalPolicy.RETAIN,
      },
    });

    this.accountId = account.getAtt("AccountId").toString();
    this.accountArn = account.getAtt("AccountArn").toString();
    this.accountName = account.getAtt("AccountName").toString();
    this.email = account.getAtt("Email").toString();

    this.resource = account;

    const tagResource = new TagResource(this, "Tags", { resource: this });
    tagResource.node.addDependency(account);
  }

  identifier(): string {
    return this.accountId;
  }

  /**
   * Enables trusted access for the AWS service (trusted service) as <strong>Delegated Administrator</strong>, which performs tasks in your organization and its accounts on your behalf.
   * @param servicePrincipal The supported AWS service that you specify
   */
  public delegateAdministrator(servicePrincipal: string) {
    const delegatedAdministrator = new DelegatedAdministrator(this, `Delegate${pascalCase(servicePrincipal)}`, {
      account: this,
      servicePrincipal: servicePrincipal,
    });
    delegatedAdministrator.node.addDependency(this.resource);
  }
}
