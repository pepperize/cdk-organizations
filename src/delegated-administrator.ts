import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId } from "aws-cdk-lib/custom-resources";
import { Construct } from "constructs";
import { IAccount } from "./account";

export interface DelegatedAdministratorProps {
  /**
   * The member account in the organization to register as a delegated administrator.
   */
  readonly account: IAccount;
  /**
   * The service principal of the AWS service for which you want to make the member account a delegated administrator.
   */
  readonly servicePrincipal: string;
}

/**
 * Enables the specified member account to administer the Organizations features of the specified AWS service. It grants read-only access to AWS Organizations service data. The account still requires IAM permissions to access and administer the AWS service.
 *
 * You can run this action only for AWS services that support this feature. For a current list of services that support it, see the column Supports Delegated Administrator in the table at AWS Services that you can use with AWS Organizations in the [AWS Organizations User Guide](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_integrate_services_list.html).
 *
 * @see https://docs.aws.amazon.com/accounts/latest/reference/using-orgs-delegated-admin.html
 */
export class DelegatedAdministrator extends Construct {
  public constructor(scope: Construct, id: string, props: DelegatedAdministratorProps) {
    super(scope, id);

    const { account, servicePrincipal } = props;

    new AwsCustomResource(this, "DelegatedAdministratorCustomResource", {
      resourceType: "Custom::Organization_DelegatedAdministrator",
      onCreate: {
        service: "Organizations",
        action: "registerDelegatedAdministrator", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#registerDelegatedAdministrator-property
        region: "us-east-1",
        physicalResourceId: PhysicalResourceId.of(`${account.accountId}:${servicePrincipal}`),
        parameters: {
          AccountId: account.accountId,
          ServicePrincipal: servicePrincipal,
        },
      },
      onDelete: {
        service: "Organizations",
        action: "registerDelegatedAdministrator", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#deregisterDelegatedAdministrator-property
        region: "us-east-1",
        parameters: {
          AccountId: account.accountId,
          ServicePrincipal: servicePrincipal,
        },
      },
      installLatestAwsSdk: false,
      policy: AwsCustomResourcePolicy.fromSdkCalls({
        resources: AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });
  }
}
