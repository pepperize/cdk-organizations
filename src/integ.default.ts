import { App, Stack } from "aws-cdk-lib";
import { Account, IamUserAccessToBilling } from "./account";
import { DelegatedAdministrator } from "./delegated-administrator";
import { EnableAwsServiceAccess } from "./enable-aws-service-access";
import { EnablePolicyType } from "./enable-policy-type";
import { FeatureSet, Organization } from "./organization";
import { OrganizationalUnit } from "./organizational-unit";
import { Policy, PolicyType } from "./policy";
import { PolicyAttachment } from "./policy-attachment";

const app = new App();
const stack = new Stack(app);

// Create an organization
const organization = new Organization(stack, "Organization", {
  featureSet: FeatureSet.ALL,
});
// Enable AWS Service Access (requires FeatureSet: ALL)
new EnableAwsServiceAccess(stack, "EnableAwsServiceAccess", {
  servicePrincipal: "service-abbreviation.amazonaws.com",
});

// Import an existing account
const account = Account.fromAccountId(stack, "ImportedAccount", {
  accountId: "123456789012",
  parent: organization.root,
});
// Enable a delegated admin account
new DelegatedAdministrator(stack, "DelegatedAdministrator", {
  account: account,
  servicePrincipal: "service-abbreviation.amazonaws.com",
});

const projects = new OrganizationalUnit(stack, "ProjectsOU", {
  organizationalUnitName: "Projects",
  parent: organization.root,
});
new Account(stack, "Project1Account", {
  accountName: "SharedAccount",
  email: "info+project1@pepperize.com",
  iamUserAccessToBilling: IamUserAccessToBilling.DENY,
  parent: projects,
});

const project2 = new OrganizationalUnit(stack, "Project2OU", {
  organizationalUnitName: "Project2",
  parent: projects,
});
new Account(stack, "Project2DevAccount", {
  accountName: "Project 2 Dev",
  email: "info+project2-dev@pepperize.com",
  parent: project2,
});
new Account(stack, "Project2ProdAccount", {
  accountName: "Project 2 Prod",
  email: "info+project2-prod@pepperize.com",
  parent: project2,
});

// Enable the service control policy (SCP) type within the organization
new EnablePolicyType(stack, "EnablePolicyType", {
  root: organization.root,
  policyType: PolicyType.SERVICE_CONTROL_POLICY,
});
// Create and attach and Service Control Policy (SCP)
const policy = new Policy(stack, "Policy", {
  content: '{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":{\\"Effect\\":\\"Allow\\",\\"Action\\":\\"s3:*\\"}}',
  description: "Enables admins of attached accounts to delegate all S3 permissions",
  policyName: "AllowAllS3Actions",
  policyType: PolicyType.SERVICE_CONTROL_POLICY,
});
new PolicyAttachment(stack, "PolicyAttachment", {
  target: organization.root,
  policy: policy,
});

export { app, stack };
