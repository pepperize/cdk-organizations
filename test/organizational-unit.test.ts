import { Template } from "@aws-cdk/assertions";
import { App, Aspects, Stack } from "@aws-cdk/core";
import { AwsSolutionsChecks } from "cdk-nag";
import { Organization, OrganizationalUnit } from "../src";

describe("Organization", () => {
  it("Should match snapshot", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "Stack");
    const organization = new Organization(stack, "Organization", {});

    // When
    new OrganizationalUnit(stack, "OrganizationalUnit", {
      organizationalUnitName: "TestOrganization",
      parent: organization.root,
    });

    // Then
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
  });

  it("Should comply to best practices", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "Stack");
    const organization = new Organization(stack, "Organization", {});

    // When
    new OrganizationalUnit(stack, "OrganizationalUnit", {
      organizationalUnitName: "TestOrganization",
      parent: organization.root,
    });

    // Then
    Aspects.of(app).add(new AwsSolutionsChecks({ verbose: true }));
  });
});
