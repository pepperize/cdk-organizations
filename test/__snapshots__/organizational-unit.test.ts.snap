// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganizationalUnit Should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "4a575666d1c2c6412590d2a56f328e040a81ad1ef59aecee31ae9b393d05f659.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "Organization06E16095": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "FeatureSet": "ALL",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "OrganizationcdkorganizationsOrganizationProviderNestedStackcdkorganizationsOrganizationProviderNestedStackResource87A8ECA1",
            "Outputs.StackOrganizationcdkorganizationsOrganizationProviderframeworkonEvent493ED7E8Arn",
          ],
        },
      },
      "Type": "Custom::Organizations_Organization",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationRootRootCustomResourceBB74F060": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "Organization06E16095",
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
      ],
      "Properties": Object {
        "Create": "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listRoots\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"responsePath\\":\\"Roots.0.Id\\"}}",
        "Delete": "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listRoots\\",\\"region\\":\\"us-east-1\\"}",
        "InstallLatestAwsSdk": false,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listRoots\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"responsePath\\":\\"Roots.0.Id\\"}}",
      },
      "Type": "Custom::Organizations_Root",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E": Object {
      "DependsOn": Array [
        "Organization06E16095",
      ],
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "organizations:ListRoots",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "OrganizationRootTagsTagResourceCBEA7B2F": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "Organization06E16095",
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "OrganizationRootRootCustomResourceBB74F060",
      ],
      "Properties": Object {
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "OrganizationRootRootCustomResourceBB74F060",
            "Roots.0.Id",
          ],
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "OrganizationRootTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource88CA0520",
            "Outputs.StackOrganizationRootTagscdkorganizationsTagResourceProviderframeworkonEvent49F750AAArn",
          ],
        },
      },
      "Type": "Custom::Organizations_TagResource",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationRootTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource88CA0520": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "Organization06E16095",
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "OrganizationRootRootCustomResourceBB74F060",
      ],
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/32735f70d37c36c06a18c3e35100d1952207019a1e1e051ea5af781463e62bce.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationalUnitOrganizationProvider9D2E0DDF": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "OrganizationRootRootCustomResourceBB74F060",
        "OrganizationRootTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource88CA0520",
        "OrganizationRootTagsTagResourceCBEA7B2F",
      ],
      "Properties": Object {
        "ImportOnDuplicate": "true",
        "Name": "TestOrganization",
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "OrganizationRootRootCustomResourceBB74F060",
            "Roots.0.Id",
          ],
        },
        "RemovalPolicy": "retain",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "OrganizationalUnitcdkorganizationsOrganizationalUnitProviderNestedStackcdkorganizationsOrganizationalUnitProviderNestedStackResource6919A668",
            "Outputs.StackOrganizationalUnitcdkorganizationsOrganizationalUnitProviderframeworkonEventD65D3462Arn",
          ],
        },
      },
      "Type": "Custom::Organizations_OrganizationalUnitProvider",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationalUnitTagsTagResource4B8852D8": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "OrganizationRootRootCustomResourceBB74F060",
        "OrganizationRootTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource88CA0520",
        "OrganizationRootTagsTagResourceCBEA7B2F",
        "OrganizationalUnitOrganizationProvider9D2E0DDF",
      ],
      "Properties": Object {
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "OrganizationalUnitOrganizationProvider9D2E0DDF",
            "Id",
          ],
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "OrganizationalUnitTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResourceF9E46EEA",
            "Outputs.StackOrganizationalUnitTagscdkorganizationsTagResourceProviderframeworkonEventF19D8AD1Arn",
          ],
        },
      },
      "Type": "Custom::Organizations_TagResource",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationalUnitTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResourceF9E46EEA": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "OrganizationRootRootCustomResourceBB74F060",
        "OrganizationRootTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource88CA0520",
        "OrganizationRootTagsTagResourceCBEA7B2F",
        "OrganizationalUnitOrganizationProvider9D2E0DDF",
      ],
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/58e821ca57f3ddf16426a6c1e372dee916f602b6515b4de7c0ad56405363fbe6.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationalUnitcdkorganizationsOrganizationalUnitProviderNestedStackcdkorganizationsOrganizationalUnitProviderNestedStackResource6919A668": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "OrganizationRootRootCustomResourceBB74F060",
        "OrganizationRootTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource88CA0520",
        "OrganizationRootTagsTagResourceCBEA7B2F",
      ],
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/3de3c4aa31376a3d15c3bb0bfcac9aaf48d478f23d1e482a1a59fb9ee3bf6854.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationcdkorganizationsOrganizationProviderNestedStackcdkorganizationsOrganizationProviderNestedStackResource87A8ECA1": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/1bc15fe5a851fabfd69a6aa5864944202df9af8b6c1c5fc95a695bf6cd411634.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
